src/conversations/Conversations.ts

```diff
@@ -97,56 +97,115 @@ export default class Conversations {
-        return undefined
+        return msg
       }
-      const peerAddress = this.getPeerAddress(msg)
-      if (
-        dmAddresses.includes(peerAddress) ||
-        peerAddress === this.client.address
-      ) {
-        // No need to update if we're already subscribed
-        return undefined
+
+      // Decode as an invite and return the envelope
+      // This gives the contentTopicUpdater everything it needs to add to the topic list
+      if (contentTopic === inviteTopic) {
+        const sealed = await SealedInvitation.fromEnvelope(env)
+        const unsealed = await sealed.v1.getInvitation(this.client.keys)
+        return ConversationV2.create(this.client, unsealed, sealed.v1.header)
+      }
+
+      const convo = convoMap.get(contentTopic)
+
+      // Decode as a V1 message if the topic matches a V1 convo
+      if (convo instanceof ConversationV1) {
+        return this.client.decodeEnvelope(env)
       }
-      dmAddresses.push(peerAddress)
-      return this.buildTopicsForAllMessages(dmAddresses, introTopic)
+
+      // Decode as a V2 message if the topic matches a V2 convo
+      if (convo instanceof ConversationV2) {
+        return convo.decodeMessage(env)
+      }
+
+      console.log('Unknown topic')
+
+      throw new Error('Unknown topic')
     }
 
-    // Filter intro topics if already streaming direct messages for that address to avoid duplicates
-    const filter = (msg: MessageV1): boolean => {
-      if (
-        msg.contentTopic === introTopic &&
-        messageHasHeaders(msg) &&
-        dmAddresses.includes(this.getPeerAddress(msg))
-      ) {
-        return false
+    const contentTopicUpdater = (msg: Conversation | Message | null) => {
+      // If we have a V1 message from the introTopic, store the conversation in our mapping
+      if (msg instanceof MessageV1 && msg.contentTopic === introTopic) {
+        const convo = new ConversationV1(
+          this.client,
+          msg.recipientAddress === this.client.address
+            ? (msg.senderAddress as string)
+            : (msg.recipientAddress as string),
+          msg.sent
+        )
+
+        if (topics.has(convo.topic)) {
+          return undefined
+        }
+
+        convoMap.set(convo.topic, convo)
+        topics.add(convo.topic)
+
+        return Array.from(topics.values())
       }
-      return true
+
+      if (msg instanceof ConversationV2) {
+        if (topics.has(msg.topic)) {
+          return undefined
+        }
+
+        convoMap.set(msg.topic, msg)
+        topics.add(msg.topic)
+
+        return Array.from(topics.values())
+      }
+
+      return undefined
     }
 
-    return Stream.create<MessageV1>(
+    const str = await Stream.create<Message | Conversation | null>(
       this.client,
-      topics,
-      async (env) => {
-        const msg = await this.client.decodeEnvelope(env)
-        return filter(msg) ? msg : undefined
-      },
+      Array.from(topics.values()),
```

Comment:
I'm not sure how to make this better, but want to point out that, by the time we subscribe to all these topics, we could have missed new intros or invites. It might be possible to subscribe to invites/intros first, then list all pre-existing, and we may have to make sure we don't subscribe to the same topic twice. Sounds messy and I really like how clean this is. Is it worth making it more complicated?

Reply:
This whole thing is full of possible race conditions.

I don't love having to do a resubscribe immediately after setting up the stream (every time we resubscribe we open up a new potential gap in the streams or potential for duplicate messages). Seems problematic no matter what.